tikv-client/src/main/java/com/pingcap/tikv/types/JsonType.java

```diff
@@ -57,171 +52,135 @@ public JsonType(MySQLType type, int flag, int len, int decimal, String charset,
+   * }
+   * }</pre>
+   *
+   * @param type type byte
+   * @param cdi codec data input
+   * @return Json element parsed
+   */
+  private JsonElement parseValue(byte type, CodecDataInput cdi) {
+    int elementCount, length;
     switch (type) {
       case TYPE_CODE_OBJECT:
-        return parseObject(di);
+        elementCount = parseUint32(cdi);
+        length = parseUint32(cdi) - 8;
+        return parseObject(cdi, elementCount, length);
       case TYPE_CODE_ARRAY:
-        return parseArray(di);
+        elementCount = parseUint32(cdi);
+        length = parseUint32(cdi) - 8;
+        return parseArray(cdi, elementCount, length);
       case TYPE_CODE_LITERAL:
-        return parseLiteralJson(di);
+        return parseLiteralJson(cdi);
       case TYPE_CODE_INT64:
-        return new JsonPrimitive(parseInt64(di));
+        return new JsonPrimitive(parseInt64(cdi));
       case TYPE_CODE_UINT64:
-        return new JsonPrimitive(parseUint64(di));
+        return new JsonPrimitive(parseUint64(cdi));
       case TYPE_CODE_FLOAT64:
-        return new JsonPrimitive(parseDouble(di));
+        return new JsonPrimitive(parseDouble(cdi));
       case TYPE_CODE_STRING:
-        long length = parseDataLength(di);
-        return new JsonPrimitive(parseString(di, length));
+        length = parseDataLength(cdi);
+        return new JsonPrimitive(parseString(cdi, length));
       default:
         throw new AssertionError("error type|type=" + (int) type);
     }
   }
 
   // * notice use this as a unsigned long
-  private long parseUint64(DataInput cdi) {
-    byte[] readBuffer = new byte[8];
-    readFully(cdi, readBuffer, 0, 8);
-
-    return ((long) (readBuffer[7]) << 56)
-        + ((long) (readBuffer[6] & 255) << 48)
-        + ((long) (readBuffer[5] & 255) << 40)
-        + ((long) (readBuffer[4] & 255) << 32)
-        + ((long) (readBuffer[3] & 255) << 24)
-        + ((readBuffer[2] & 255) << 16)
-        + ((readBuffer[1] & 255) << 8)
-        + ((readBuffer[0] & 255) << 0);
+  private long parseUint64(CodecDataInput cdi) {
+    return cdi.readLongLSB();
   }
 
-  private void readFully(DataInput cdi, byte[] readBuffer, final int off, final int len) {
-    try {
-      cdi.readFully(readBuffer, off, len);
-    } catch (IOException e) {
-      throw new UncheckedIOException(e);
-    }
+  private long parseInt64(CodecDataInput cdi) {
+    return cdi.readLongLSB();
   }
 
-  private long parseInt64(DataInput cdi) {
-    byte[] readBuffer = new byte[8];
-    readFully(cdi, readBuffer);
-    return ((long) readBuffer[7] << 56)
-        + ((long) (readBuffer[6] & 255) << 48)
-        + ((long) (readBuffer[5] & 255) << 40)
-        + ((long) (readBuffer[4] & 255) << 32)
-        + ((long) (readBuffer[3] & 255) << 24)
-        + ((readBuffer[2] & 255) << 16)
-        + ((readBuffer[1] & 255) << 8)
-        + ((readBuffer[0] & 255) << 0);
+  private int parseUint32(CodecDataInput cdi) {
```

Comment:
s/parseUint32/parseUint32LSB ?

Reply:
I don't want to use LSB here because the function here does not need to be aware of its implementation.